Source/AuthenticationInterceptor.swift

```diff
@@ -0,0 +1,398 @@
+    /// matches the values in the `Credential`. In the case of OAuth2, this would mean that the Bearer token in the
+    /// `Authorization` header of the `URLRequest` matches the access token in the `Credential`. If it matches, then we
+    /// know the `Credential` was used to authenticate the `URLRequest` and should return `true`. If the Bearer token
+    /// did not match the access token, then you should return `false`.
+    ///
+    /// - Parameters:
+    ///   - urlRequest: The `URLRequest`.
+    ///   - credential: The `Credential`.
+    ///
+    /// - Returns: `true` if the `URLRequest` is authenticated with the `Credential`, `false` otherwise.
+    func isRequest(_ urlRequest: URLRequest, authenticatedWith credential: Credential) -> Bool
+}
+
+// MARK: -
+
+/// Represents various authentication failures that fail when using the `AuthenticationInterceptor`. All errors are
+/// still vended from Alamofire as `AFError` types. The `AuthenticationError` instances will be embedded within
+/// `AFError` `.requestAdaptationFailed` or `.requestRetryFailed` cases.
+public enum AuthenticationError: Error {
+    /// The credential was missing so the request could not be authenticated.
+    case missingCredential
+
+    /// The credential was refreshed too many times within the `RefreshWindow`.
+    case excessiveRefresh
+}
+
+// MARK: -
+
+/// The `AuthenticationInterceptor` class manages the queuing and threading complexity of authenticating requests.
+/// It relies on an `Authenticator` type to handle the actual `URLRequest` authentication and `Credential` refresh.
+public class AuthenticationInterceptor<AuthenticatorType>: RequestInterceptor where AuthenticatorType: Authenticator {
+
+    // MARK: Typealiases
+
+    /// Represents the type of credential used to authenticate requests.
+    public typealias Credential = AuthenticatorType.Credential
+
+    // MARK: Helper Types
+
+    /// A container type that defines a refresh window to use to identify excessive refresh calls. When enabled, prior
+    /// to executing a refresh, the `AuthenticationInterceptor` compares the timestamp history of previous refresh calls
+    /// against the `RefreshWindow`. If more refreshes have occurred within the refresh window than allowed, the refresh
+    /// is cancelled an an `AuthorizationError.excessiveRefresh` error is thrown.
+    public struct RefreshWindow {
+        /// The past time interval to inspect to see how many refreshes occurred.
+        public let interval: TimeInterval
+
+        /// The total refresh attempts allowed within the refresh window before throwing an `.excessiveRefresh` error.
+        public let maximumAttempts: Int
+
+        /// Creates a `RefreshWindow` instance from the specified `interval` and `maximumAttempts`.
+        ///
+        /// - Parameters:
+        ///   - interval:        The past `TimeInterval` to inspect.
```

Comment:
This should really match the property description but I can't think of a more succinct way to express my update up there.

Reply:
I updated based on my new edits in cf6f5c2d.